### PR TITLE
Feature/shipyard ready registry v2

### PR DESCRIPTION
**Pull request by borzamircea**
Added support for registry v2 with authentification

### CODE REVIEW

File: auth/ldap/ldap.go [hunk 1]
```diff
@@ -45,13 +44,6 @@ func (a LdapAuthenticator) Authenticate(username, password, hash string) (bool,
...
 
 	dn := fmt.Sprintf("cn=%s,%s", username, a.BaseDN)
-
-	if strings.Contains(a.BaseDN, "{username}") {
-		dn = strings.Replace(a.BaseDN, "{username}", username, -1)
-	}
-
-	log.Debugf("ldap authentication: dn=%s", dn)
-
 	if err := l.Bind(dn, password); err != nil {
```
Comment (ehazlett) [inline auth/ldap/ldap.go:48]: Why are these removed?

File: controller/manager/manager.go [hunk 2]
```diff
@@ -772,12 +814,36 @@ func (m DefaultManager) Registry(name string) (*shipyard.Registry, error) {
...
+}
+
+func (m DefaultManager) RegistryByAddress(addr string) (*shipyard.Registry, error) {
+	res, err := r.Table(tblNameRegistries).Filter(map[string]string{"addr": addr}).Run(m.session)
+	defer res.Close()
 	if err != nil {
 		return nil, err
 	}
+	if res.IsNil() {
+		log.Debugf("its nil!! it found nothing")
```
Comment (ehazlett) [inline controller/manager/manager.go]: can this be removed or more descriptive?  this by itself in the logs is not very helpful.

File: controller/manager/manager.go [hunk 3]
```diff
@@ -772,12 +814,36 @@ func (m DefaultManager) Registry(name string) (*shipyard.Registry, error) {
...
+		return nil, ErrRegistryDoesNotExist
+	}
+	var reg *shipyard.Registry
+	if err := res.One(&reg); err != nil {
+		log.Debugf("problem with res.One")
```
Comment (ehazlett) [inline controller/manager/manager.go]: same as above comment (remove or more descriptive debug)

File: controller/manager/manager.go [hunk 4]
```diff
@@ -772,12 +814,36 @@ func (m DefaultManager) Registry(name string) (*shipyard.Registry, error) {
...
+		return nil, err
+	}
+
+	if err := reg.InitRegistryClient(); err != nil {
+		log.Errorf("%s", err.Error())
```
Comment (ehazlett) [inline controller/manager/manager.go]: you can just `log.Error(err)`

File: registry/v2/registry.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+	}, nil
+}
+
+func (client *RegistryClient) doRequest(method string, path string, body []byte, headers map[string]string) ([]byte, http.Header, error) {
+	b := bytes.NewBuffer(body)
+
+	req, err := http.NewRequest(method, client.URL.String()+"/v2"+path, b)
+	log.Debugf("Method: %s   URL: %s", method, client.URL.String()+"/v2"+path)
+	if err != nil {
+		log.Debugf("Error on doRequest")
```
Comment (ehazlett) [inline registry/v2/registry.go]: Same as comments above.  This is not very helpful by itself.

File: registry/v2/registry.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+			req.Header.Add(header, value)
+		}
+	}
+
+	resp, err := client.httpClient.Do(req)
+	if err != nil {
+		if !strings.Contains(err.Error(), "connection refused") && client.tlsConfig == nil {
+			return nil, nil, fmt.Errorf("%s. Are you trying to connect to a TLS-enabled daemon without TLS?", err)
+		}
+		log.Debugf("Connection refused")
```
Comment (ehazlett) [inline registry/v2/registry.go]: remove

File: registry/v2/registry.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+		return nil, nil, err
+	}
+
+	defer resp.Body.Close()
+
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Debugf("Error on ioutil.ReadAll")
```
Comment (ehazlett) [inline registry/v2/registry.go]: remove

File: registry/v2/registry.go [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+		return nil, nil, err
+	}
+
+	if resp.StatusCode == 404 {
+		log.Debugf("Error on resp.StatusCode == 404")
```
Comment (ehazlett) [inline registry/v2/registry.go]: remove

File: registry/v2/registry.go [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+		return nil, nil, ErrNotFound
+	}
+
+	if resp.StatusCode >= 400 {
+		log.Debugf("Error on resp.StatusCode >= 400")
```
Comment (ehazlett) [inline registry/v2/registry.go]: remove

File: registry/v2/registry.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+func (client *RegistryClient) Search(query string) ([]*Repository, error) {
+	type repo struct {
+		Repositories []string `json:"repositories"`
+	}
+
+	uri := fmt.Sprintf("/_catalog")
+	data, _, err := client.doRequest("GET", uri, nil, nil)
+	if err != nil {
+		log.Debugf("Error on client.doRequest(GET, uri, nil, nil)")
```
Comment (ehazlett) [inline registry/v2/registry.go]: more descriptive / remove

File: registry/v2/registry.go [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+		return nil, err
+	}
+
+	log.Debugf("Response from /_catalog %s", string(data))
+
+	res := &repo{}
+	if err := json.Unmarshal(data, &res); err != nil {
+		log.Debugf("Error on json.Unmarshal(data, &res)")
```
Comment (ehazlett) [inline registry/v2/registry.go]: remove or more descriptive (if more descriptive fix the `Debugf` -- you aren't passing anything to format)

File: registry/v2/registry.go [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+
+	repos := []*Repository{}
+
+	// simple filter for list
+	for _, k := range res.Repositories {
+		if strings.Index(k, query) == 0 {
+			tl, err := client.getTags(k)
+			if err != nil {
+				msg := fmt.Sprintf("Error on getting tags: %s", err.Error())
+				log.Error(msg)
```
Comment (ehazlett) [inline registry/v2/registry.go]: collapse to `log.Errorf("error getting tags: %s", err)`

File: utils/emitter/emitter.go [hunk 13]
```diff
@@ -0,0 +1,65 @@
+package emitter
```
Comment (ehazlett) [inline utils/emitter/emitter.go]: I'm not seeing where this is used anywhere.  Can it be removed?

File: utils/tlsutils/tlsutils.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+	// server cert
+	keypair, err := tls.X509KeyPair(serverCert, serverKey)
+	if err != nil {
+		return &tlsConfig, err
+
+	}
+	tlsConfig.Certificates = []tls.Certificate{keypair}
+
+	if allowInsecure {
+		tlsConfig.InsecureSkipVerify = true
```
Comment (ehazlett) [inline utils/tlsutils/tlsutils.go]: this appears to be set twice (line 55)

File: utils/tlsutils/tlsutils.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+	}
+
+	return &tlsConfig, nil
+}
+
+func newCertificate(org string) (*x509.Certificate, error) {
+	now := time.Now()
+	// need to set notBefore slightly in the past to account for time
+	// skew in the VMs otherwise the certs sometimes are not yet valid
+	notBefore := time.Date(now.Year(), now.Month(), now.Day(), now.Hour(), now.Minute()-5, 0, 0, time.Local)
```
Comment (ehazlett) [inline utils/tlsutils/tlsutils.go]: what about `notBefore := time.Now().Add(-time.Second*300)` or something.

**Comment by ehazlett** [comment]
Please check all of the `log.Error` and `log.Debug`.  I believe we log the error on return so we are probaly getting double logs with this.

**Comment by ehazlett** [comment]
Looks pretty good!  Thanks for opening!

**Comment by borzamircea** [comment]
Hi @ehazlett . Thanks for the review. I just pushed the changes you requested. Thanks!

**Comment by soninob** [comment]
Hey @ehazlett, any progress for this merge?

**Comment by ehazlett** [comment]
I'll get this tested.  Thanks!

**Comment by soninob** [comment]
Hey @ehazlett, 

Any news about the testing?
Waiting for your feedbacks.

**Comment by wohckcin** [comment]
Hey @ehazlett, 
totally loving shipyard, please keep doing your excellent works. The only downside about shipyard I can think of is not supporting Registry v2. Really hope this feature could be merged into the master in the near future. Cheers!

**Comment by jesuscoz** [comment]
Changes for  **auth/ldap/ldap.go** breaks ldap {username} substitution, this code should be placed before ldap bind invocation, otherwise return false,
Cheers

**Comment by timofurrer** [comment]
Awesome work, guys! Any progress with testing?

**Comment by ehazlett** [comment]
Looks good!  What is the "remarks" field in the table?  Also, the size is not being calculated.

2016-09-19-105953_2509x298_scrot

File: registry/v2/registry.go [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+		Name:         name,
+		Tag:          tag,
+		HasProblems:  true,
+		Architecture: "",
+		RegistryUrl:  registryUrl,
+	}
+
+	uri := fmt.Sprintf("/%s/manifests/%s", name, tag)
+
+	log.Infof("requesting manifest for %s", uri)
```
Comment (ehazlett) [inline registry/v2/registry.go]: This should be debug.

File: registry/v2/registry.go [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+	repo := &Repository{}
+	if err := json.Unmarshal(data, &repo); err != nil {
+		invalidRepository.Message = fmt.Sprintf("Error when binding manifests for %s, error = %s", uri, err.Error())
+		log.Error(invalidRepository.Message)
+		return invalidRepository, err
+	}
+
+	repo.RegistryUrl = registryUrl
+	repo.Digest = hdr.Get("Docker-Content-Digest")
+	log.Infof("Got docker content digest %s", repo.Digest)
```
Comment (ehazlett) [inline registry/v2/registry.go]: This should be debug.

**Comment by ehazlett** [comment]
Ok after addressing these we should be good to merge.  Functionality looked good; couple of nits for logging to reduce chatter.  I'm not sure why we need the "Remarks" column and we need to get the size calculating.  Other than that, it looks good.  Great job!

**Comment by cysnake4713** [comment]
+1 hope this one merge asap~

**Comment by soninob** [comment]
@ehazlett the Remarks field is being populated in case we can't get the manifest properly.
We are working on getting the size, we need to get a different manifest with a different Header, hopefully, today it will be ready.

**Comment by AndreeaDora** [comment]
Hey @ehazlett . Thank you for the review. I added the changes so we will display the size for a repository, and the changes for the logging.

**Comment by borzamircea** [comment]
Great job @AndreeaDora !

**Comment by joshpelz** [comment]
+1 
eagerly awaiting this functionality

**Review by vroomvroomvroom — APPROVED** [review]
very cool!

**Comment by vroomvroomvroom** [comment]
appears all checks have passed! @ehazlett when will PR get merged?

**Comment by ehazlett** [comment]
@AndreeaDora thanks!  i'll take a look.

**Comment by ehazlett** [comment]
I'm unable to ping registries now.  I set up a local registry, pushed an image and tried to add it to Shipyard.  I receive "cannot ping".  However, I can `curl` it from the shipyard controller via an exec:

2016-10-05-100525_647x957_scrot

```
/ # curl -v registry:5000
* Rebuilt URL to: registry:5000/
*   Trying 172.17.0.2...
* TCP_NODELAY set
* Connected to registry (172.17.0.2) port 5000 (#0)
> GET / HTTP/1.1
> Host: registry:5000
> User-Agent: curl/7.50.3
> Accept: */*
> 
< HTTP/1.1 200 OK
< Cache-Control: no-cache
< Date: Wed, 05 Oct 2016 14:03:55 GMT
< Content-Length: 0
< Content-Type: text/plain; charset=utf-8
< 
* Curl_http_done: called premature == 0
* Connection #0 to host registry left intact
```

**Comment by ehazlett** [comment]
actually wait -- this may be a DNS issue on my side; checking

**Comment by ehazlett** [comment]
ok this was a DNS issue on my part.  

Can someone explain the "Remarks" field -- I'm not sure what it's used for.

**Comment by ehazlett** [comment]
2016-10-05-101205_594x248_scrot

**Comment by ehazlett** [comment]
ah i see above it's for errors.  i think this is fine.  lgtm

**Comment by eitansch** [comment]
Thanks @ehazlett! great work everyone.

**Comment by vroomvroomvroom** [comment]
awesome!  pulling in AM  thanks to all for the great work

**Comment by ph1s0** [comment]
Hey I just noticed that the button for deleting an image from a registry has been removed. Is there any reason for that function not beeing implemented for registry v2?
image

Also I had some issues when trying to use shipyard with a v1 registry. It seems shipyard is not able to ping any v1 registry.

**Comment by ogerasymenko** [comment]
@ehazlett Hi. Earlier at https://github.com/shipyard/shipyard/pull/852#issuecomment-251684627 you mention what receive "cannot ping" registry.  Now I try to play with Shipyard and have same situation. But not sure this is related with DNS. In registry container have next logs:
http: TLS handshake error from 172.16.0.100:56324: tls: first record does not look like a TLS handshake
I create local registry:2 with self-signed certificates. Maybe this related with SSL settings?
